src/NHibernate/IFutureValueAsync.cs

```diff
@@ -0,0 +1,9 @@
+using System.Threading.Tasks;
+
+namespace NHibernate
+{
+	public interface IFutureValueAsync<T>
+	{
+		Task<T> GetValue();
```

Comment:
Great, to me it looks a lot more nicer. I think this could be simplified even more by getting rid of the `IQuery.ToFutureAsync` too, which is indeed a synchronous method returning an async enabled object. A single `ToFuture` returning an interface capable of both would be nice. But we would have to introduce something like a `IFutureValue<T>` for this, deriving from both `IEnumerable` and `IAsyncEnumerable`. And that would be a possible breaking change, although I do not think much people have implemented their own NHibernate query API.